13724 console parallelization

mattamon opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.6`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.6` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #13724 

## Additional info  
Added command options, removed deprecated methods and added deprecation warning for configureParallelization

mattamon committed:
Adding options (main-process, segment-size)

mattamon committed:
Added input options and removed deprecated methods
Added a deprecation warning for configureParallelization

jdreesen commented on lib/Console/Traits/Parallelization.php:
```diff
@@ -33,6 +33,9 @@ trait Parallelization
     use ParallelizationBase;
 
+    /**
+     * @deprecated Deprecated since 2.0.0 and will be removed in 3.0.0. Use ParallelizationInput::configureCommand() instead.
```
Version `2.0.0` and `3.0.0` of what?

mattamon replied:
Added the bundle name to make it more clear.

mattamon committed:
Add bundle name to deprecation note

brusch commented on lib/Console/Traits/Parallelization.php:
```diff
@@ -33,6 +33,10 @@ trait Parallelization
     use ParallelizationBase;
 
+    /**
+     * @deprecated Deprecated since webmozarts/console-parallelization 2.0.0 and will be removed in webmozarts/console-parallelization 3.0.0. Use ParallelizationInput::configureCommand() instead.
```
@mattamon Shouldn't we directly implement `configureCommand()`?

mattamon replied:
Yes I thought about that too, but the configureCommand is not part of the  \Webmozarts\Console\Parallelization\Parallelization Trait anymore, it is now it's own class, so I am not sure how to handle this.
https://github.com/webmozarts/console-parallelization/blob/main/src/Input/ParallelizationInput.php

brusch replied:
@mattamon you can use that probably in the `configure()` method ...?

mattamon replied:
I saw that comment too, but I am not sure about it since there is no configure method in the trait.

brusch replied:
The `configure()` method you implement in the command, and there you call the method from the trait to configure the command, see e.g.: 
https://github.com/pimcore/pimcore/blob/11.x/bundles/CoreBundle/src/Command/CacheClearCommand.php#L43

mattamon committed:
Removing configureParallelization method.
Using configureCommand Webmozarts\Console\Parallelization\Input\ParallelizationInput configureCommand directly
Breaking change since the p shortcut is not available anymore

mattamon committed:
Adding upgrade notes

dvesh3 commented on lib/Console/Traits/ParallelizationBase.php:
```diff
@@ -16,17 +16,25 @@
             \Webmozarts\Console\Parallelization\Parallelization::execute as parentExecute;
         }
+
+        protected function configureCommand(Command $command): void
```
why not keep it static? as the called method from Webmozarts is also static.

mattamon replied:
Should we also make it public then?

dvesh3 replied:
`protected` is fine, since only inheriting commands will call this method.

dvesh3 commented on lib/Console/Traits/ParallelizationBase.php:
```diff
@@ -58,5 +66,47 @@ public function execute(InputInterface $input, OutputInterface $output): int
+
+        protected function configureCommand(Command $command): void
+        {
+            $command
```
No need to extend options here as the parallelization is disabled.

mattamon committed:
Keep methods static
Removing extending options

dvesh3 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
```diff
@@ -67,7 +67,9 @@ Please make sure to set your preferred storage location ***before*** migration.
 - [Document Editables] Removed method_exists bc layer for `getDataEditmode()`, `rewriteIds()` & `load()`, please use the corresponding interfaces `EditmodeDataInterface`, `IdRewriterInterface` & `LazyLoadingInterface` instead.
-- [Parallelization] Removed `webmozarts/console-parallelization` dependency due to maintainability reasons. If you still want to use parallelization for console commands, please add the dependency to your own `composer.json`.
+- [Parallelization] 
+  - Removed `webmozarts/console-parallelization` dependency due to maintainability reasons. If you still want to use parallelization for console commands, please add the dependency to your own `composer.json`.
```
```suggestion
  - Removed `webmozarts/console-parallelization` dependency to make parallelization optional. If you still want to use parallelization for console commands, please add the dependency to your own `composer.json`.
```

dvesh3 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
```diff
@@ -67,7 +67,9 @@ Please make sure to set your preferred storage location ***before*** migration.
-- [Parallelization] Removed `webmozarts/console-parallelization` dependency due to maintainability reasons. If you still want to use parallelization for console commands, please add the dependency to your own `composer.json`.
+- [Parallelization] 
+  - Removed `webmozarts/console-parallelization` dependency due to maintainability reasons. If you still want to use parallelization for console commands, please add the dependency to your own `composer.json`.
+  - Removing the `configureParallelization()` method from the trait. Please call `$this->configureCommand()` instead to use the standard `webmozarts/console-parallelization` functionality
```
```suggestion
  - Removed the deprecated `Parallelization::configureParallelization()` method.
```
New method should be added on 10.6 with deprecation notes.

mattamon committed:
Update doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md

Co-authored-by: Divesh Pahuja <divesh.pahuja@pimcore.com>

mattamon committed:
Update doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md

Co-authored-by: Divesh Pahuja <divesh.pahuja@pimcore.com>

dvesh3 commented on bundles/EcommerceFrameworkBundle/src/Command/IndexService/ProcessPreparationQueueCommand.php:
```diff
@@ -62,7 +62,7 @@ protected function configure()
         parent::configure();
 
-        self::configureParallelization($this);
+        $this->configureCommand($this);
```
```suggestion
        self::configureCommand($this);
```

dvesh3 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
```diff
@@ -67,7 +67,9 @@ Please make sure to set your preferred storage location ***before*** migration.
 - [Elements] Changed method signature on `Pimcore\Model\Element\ElementInterface::save()`, this changes the `::save()` method on all classes (e.g. DataObjects and Pages) implementing the interface, including those inheriting from `Concrete`/`AbstractObject`, see [#13207](https://github.com/pimcore/pimcore/issues/13207)
 - [Document Editables] Removed method_exists bc layer for `getDataEditmode()`, `rewriteIds()` & `load()`, please use the corresponding interfaces `EditmodeDataInterface`, `IdRewriterInterface` & `LazyLoadingInterface` instead.
-- [Parallelization] Removed `webmozarts/console-parallelization` dependency due to maintainability reasons. If you still want to use parallelization for console commands, please add the dependency to your own `composer.json`.
+- [Parallelization] 
```
```suggestion
- [Commands] 
```

mattamon committed:
Changed configureCommand to static call
Updated upgrade notes

dvesh3 committed:
Merge branch '11.x' into 13724-console-parallelization
